### DISCUSSION TITLE
Support for DDS on_inconsistent_topic topic listener callback

### DISCUSSION THREAD

**Discussion by clalancette** [eProsima/Fast-DDS#3168] [Ideas]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

DDS 1.4, 2.2.4.3 allows for an "inconsistent topic" callback to be called when a topic with the same name but different characteristics is detected.

In Fast-DDS, this callback can be set using the Topic Listener class https://github.com/eProsima/Fast-DDS/blob/master/include/fastdds/dds/topic/TopicListener.hpp .  However, nothing in the implementation will ever call that callback.  That was determined both through code inspection, and also through empirical testing using https://github.com/clalancette/dds_experimentation/tree/clalancette/cleanups/fastdds/HelloWorldExample .

It would be great if Fast-DDS would finish up support for this so we can get these callbacks at the ROS 2 layer.

### Current behavior

The `on_inconsistent_topic` callback can be registered, but will never be called.

### Steps to reproduce

1.  Clone https://github.com/clalancette/dds_experimentation/tree/clalancette/cleanups/fastdds/HelloWorldExample (first time)
2.  Build the example with `mkdir build ; cd build ; cmake .. -DV1TYPE=T1 -DV2TYPE=T1 -DV3TYPE=T1 ; make`
3.  Run the Subscriber: `./HelloWorldSubscriber v1`
4.  Clone https://github.com/clalancette/dds_experimentation/tree/clalancette/cleanups/fastdds/HelloWorldExample (second time)
5.  Build the example with `mkdir build ; cd build ; cmake .. -DV1TYPE=T3 -DV2TYPE=T1 -DV3TYPE=T1 ; make`
6.  Run the Publisher: `./HelloWorldPublisher v1`

### Fast DDS version/commit

Commit bffc8768e34c211a2024241051791214f370324d

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

UDPv4

### Additional context

Related to https://github.com/ros2/ros2/issues/1361

**Comment by EduPonz** [comment]
Hi @clalancette , as documented [here](https://fast-dds.docs.eprosima.com/en/v2.9.0/fastdds/dds_layer/topic/topicListener/topicListener.html#topiclistener), as of Fast DDS v2.9.0, `on_inconsistent_topic` functionality is not yet implemented. Therefore, I'm converting this issue into a discussion on the channel appropriate for feature requests. We will plan internally how to support this feature in Fast DDS for future releases.

Thanks!

**Comment by clalancette** [comment]
Related to https://github.com/ros2/ros2/issues/1361

**Comment by MiguelCompany** [comment]
@clalancette I have a PoC you can test on [this branch](https://github.com/eProsima/Fast-DDS/tree/feature/inconsistent_topic_status/poc)

**Reply by clalancette** [reply]
I gave it a quick whirl, but I think Fast-DDS is doing something different than Connext here.

In particular, what I did was to build a ROS 2 workspace twice; once unmodified, and once with the `effort` field of https://github.com/ros2/demos/blob/rolling/pendulum_msgs/msg/JointState.msg changed to a `string`.  Then I ran the following under Connext:

Terminal 1:
```
RMW_IMPLEMENTATION=rmw_connextdds ros2 topic pub /chatter pendulum_msgs/msg/JointState '{position: 0, velocity: 0, effort: 0}'
publisher: beginning loop
publishing #1: pendulum_msgs.msg.JointState(position=0.0, velocity=0.0, effort=0.0)
...
```

Terminal 2:
```
RMW_IMPLEMENTATION=rmw_connextdds ros2 topic echo /chatter pendulum_msgs/msg/JointState
[WARN] [1678975432.330708363] [_ros2cli_3752035]: Incompatible type on topic '/chatter', no messages will be sent to it.
```

That's what I expect to happen.

I then rebuilt the workspaces with the `feature/inconsistent_topic_status/poc` branch for Fast-DDS, and it is still transferring data as far as I can tell:

Terminal 1:
```
RMW_IMPLEMENTATION=rmw_fastrtps_cpp ros2 topic pub /chatter pendulum_msgs/msg/JointState '{position: 0, velocity: 0, effort: 0}'
publisher: beginning loop
publishing #1: pendulum_msgs.msg.JointState(position=0.0, velocity=0.0, effort=0.0)
...
```

Terminal 2:
```
RMW_IMPLEMENTATION=rmw_fastrtps_cpp ros2 topic echo /chatter pendulum_msgs/msg/JointState
position: 0.0
velocity: 0.0
effort: ''
---
...
```

Any thoughts on what might be going on here?

**Comment by MiguelCompany** [comment]
@clalancette Fast DDS is not sending type information on the discovery by default (yet), so you should check with different type names.

**Reply by clalancette** [reply]
> @clalancette Fast DDS is not sending type information on the discovery by default (yet), so you should check with different type names.

Ah, thanks for the hint.  Indeed, with using different type names, this works all the way through `rmw_fastrtps_cpp` and all the way to `rclpy`.  Great works, thanks for this.
